File: miscellaneous/graph-dfs/20221220/graph-dfs.php
<?php

declare(strict_types=1);

function graphDfs(array &$graph): array
{
    $values = [];
    dfs(graph: $graph, values: $values, vertex: 0);
    return $values;
}

function dfs(array &$graph, &$values, $vertex): void
{
    $values[] = $vertex;

    foreach ($graph[$vertex] as $neighbor) {
        if (!in_array($neighbor, $values)) {
            dfs($graph, $values, $neighbor);
        }
    }
}

File: miscellaneous/graph-dfs/20221220/graph_dfs.py
from typing import List


def graph_dfs(graph: [List[List[int]]]) -> List[int]:
    values: List[int] = []
    dfs(graph, values, 0)
    return values


def dfs(graph: List[List[int]], values: List[int], vertex: int) -> None:
    values.append(vertex)

    for neighbor in graph[vertex]:
        if neighbor not in values:
            dfs(graph, values, neighbor)


